File: algorithms/divide-and-conquer/strassens-matrix/kotlin/StrassensMatrix.kt
fun strassensMatrix(arr: IntArray): IntArray {
    val n = arr[0]
    var sz = 1
    while (sz < n) sz *= 2

    fun makeMat(s: Int) = Array(s) { IntArray(s) }
    fun subM(m: Array<IntArray>, r: Int, c: Int, s: Int): Array<IntArray> {
        val res = makeMat(s)
        for (i in 0 until s) for (j in 0 until s) res[i][j] = m[r+i][c+j]
        return res
    }
    fun addM(a: Array<IntArray>, b: Array<IntArray>, s: Int): Array<IntArray> {
        val r = makeMat(s)
        for (i in 0 until s) for (j in 0 until s) r[i][j] = a[i][j] + b[i][j]
        return r
    }
    fun subM2(a: Array<IntArray>, b: Array<IntArray>, s: Int): Array<IntArray> {
        val r = makeMat(s)
        for (i in 0 until s) for (j in 0 until s) r[i][j] = a[i][j] - b[i][j]
        return r
    }

    fun mul(a: Array<IntArray>, b: Array<IntArray>, s: Int): Array<IntArray> {
        val c = makeMat(s)
        if (s == 1) { c[0][0] = a[0][0] * b[0][0]; return c }
        val h = s / 2
        val a11=subM(a,0,0,h); val a12=subM(a,0,h,h)
        val a21=subM(a,h,0,h); val a22=subM(a,h,h,h)
        val b11=subM(b,0,0,h); val b12=subM(b,0,h,h)
        val b21=subM(b,h,0,h); val b22=subM(b,h,h,h)
        val m1=mul(addM(a11,a22,h),addM(b11,b22,h),h)
        val m2=mul(addM(a21,a22,h),b11,h)
        val m3=mul(a11,subM2(b12,b22,h),h)
        val m4=mul(a22,subM2(b21,b11,h),h)
        val m5=mul(addM(a11,a12,h),b22,h)
        val m6=mul(subM2(a21,a11,h),addM(b11,b12,h),h)
        val m7=mul(subM2(a12,a22,h),addM(b21,b22,h),h)
        val c11=addM(subM2(addM(m1,m4,h),m5,h),m7,h)
        val c12=addM(m3,m5,h)
        val c21=addM(m2,m4,h)
        val c22=addM(subM2(addM(m1,m3,h),m2,h),m6,h)
        for (i in 0 until h) for (j in 0 until h) {
            c[i][j]=c11[i][j]; c[i][j+h]=c12[i][j]
            c[i+h][j]=c21[i][j]; c[i+h][j+h]=c22[i][j]
        }
        return c
    }

    val a = makeMat(sz); val b = makeMat(sz)
    for (i in 0 until n) for (j in 0 until n) {
        a[i][j] = arr[1+i*n+j]; b[i][j] = arr[1+n*n+i*n+j]
    }
    val result = mul(a, b, sz)
    return IntArray(n * n) { result[it / n][it % n] }
}

File: algorithms/divide-and-conquer/strassens-matrix/rust/strassens_matrix.rs
type Mat = Vec<Vec<i32>>;

fn make_mat(n: usize) -> Mat {
    vec![vec![0i32; n]; n]
}

fn sub_mat(m: &Mat, r: usize, c: usize, sz: usize) -> Mat {
    let mut res = make_mat(sz);
    for i in 0..sz { for j in 0..sz { res[i][j] = m[r+i][c+j]; } }
    res
}

fn add_mat(a: &Mat, b: &Mat, sz: usize) -> Mat {
    let mut r = make_mat(sz);
    for i in 0..sz { for j in 0..sz { r[i][j] = a[i][j] + b[i][j]; } }
    r
}

fn sub_mat2(a: &Mat, b: &Mat, sz: usize) -> Mat {
    let mut r = make_mat(sz);
    for i in 0..sz { for j in 0..sz { r[i][j] = a[i][j] - b[i][j]; } }
    r
}

fn multiply(a: &Mat, b: &Mat, n: usize) -> Mat {
    let mut c = make_mat(n);
    if n == 1 { c[0][0] = a[0][0] * b[0][0]; return c; }
    let h = n / 2;
    let (a11, a12) = (sub_mat(a,0,0,h), sub_mat(a,0,h,h));
    let (a21, a22) = (sub_mat(a,h,0,h), sub_mat(a,h,h,h));
    let (b11, b12) = (sub_mat(b,0,0,h), sub_mat(b,0,h,h));
    let (b21, b22) = (sub_mat(b,h,0,h), sub_mat(b,h,h,h));

    let m1 = multiply(&add_mat(&a11,&a22,h), &add_mat(&b11,&b22,h), h);
    let m2 = multiply(&add_mat(&a21,&a22,h), &b11, h);
    let m3 = multiply(&a11, &sub_mat2(&b12,&b22,h), h);
    let m4 = multiply(&a22, &sub_mat2(&b21,&b11,h), h);
    let m5 = multiply(&add_mat(&a11,&a12,h), &b22, h);
    let m6 = multiply(&sub_mat2(&a21,&a11,h), &add_mat(&b11,&b12,h), h);
    let m7 = multiply(&sub_mat2(&a12,&a22,h), &add_mat(&b21,&b22,h), h);

    let c11 = add_mat(&sub_mat2(&add_mat(&m1,&m4,h),&m5,h),&m7,h);
    let c12 = add_mat(&m3,&m5,h);
    let c21 = add_mat(&m2,&m4,h);
    let c22 = add_mat(&sub_mat2(&add_mat(&m1,&m3,h),&m2,h),&m6,h);

    for i in 0..h { for j in 0..h {
        c[i][j]=c11[i][j]; c[i][j+h]=c12[i][j];
        c[i+h][j]=c21[i][j]; c[i+h][j+h]=c22[i][j];
    }}
    c
}

pub fn strassens_matrix(arr: &[i32]) -> Vec<i32> {
    let n = arr[0] as usize;
    let mut sz = 1;
    while sz < n { sz *= 2; }
    let mut a = make_mat(sz);
    let mut b = make_mat(sz);
    for i in 0..n { for j in 0..n {
        a[i][j] = arr[1+i*n+j];
        b[i][j] = arr[1+n*n+i*n+j];
    }}
    let result = multiply(&a, &b, sz);
    let mut out = Vec::new();
    for i in 0..n { for j in 0..n { out.push(result[i][j]); } }
    out
}
